draft: 422 error when trying to send data to /users/{id}

Sebastian-Wlo opened the pull request:
I'm trying to figure out what is wrong with this submission form. As far as I understand, `/users/{id}` should accept a 'PATCH' request with at least the 'name' value.
From what I can see in the Network Monitor, that is what I'm trying to send.

The initial idea I had for the page was to create separate forms for editing name, password, and adding admin privileges, but I don't understand why I'm getting this error there.

This is the request payload as it appears in the console:
```bash
------geckoformboundaryac2e682e793f835d60c4cbabb08c71ad
Content-Disposition: form-data; name="name"

admin0
------geckoformboundaryac2e682e793f835d60c4cbabb08c71ad--
```

<!--If pull request closes an issue, write its number in the place of XXXXXX.-->

It doesn't seem different from what `login` and `register` forms generate.

<!--Closes #166-->

gikf commented:
Oh, and regarding the password change... we should think about something to make sure it's being changed by somebody knowing the current password.

Sebastian-Wlo commented on frontend/src/pages/MeEditPage.jsx:
```diff
@@ -1,10 +1,227 @@
+                <UserIcon />
+                <input
+                  id='name'
+                  {...register('name')}
```
Fair point - especially since the default value is already the current name. I'll add it in a minute.

gikf commented:
> Yeah, that would probably be a good idea - how much work would it take to make the `/me/edit` API path run a check against the current hashed password? Is this even possible to do on an 'optional' value?

Not much, couple lines. I'm fairly sure there's way to make that validation _automatic_ via model, but adding _manual_ check in the route function wouldn't be troublesome.

Sebastian-Wlo commented on frontend/src/pages/MeEditPage.jsx:
```diff
@@ -1,10 +1,227 @@
+
+  const onSubmit = async formData => {
+    try {
+      await fetchFromApi(`/users/${data.id}`, {
```
I changed one in `useEffect()` earlier and missed that one, thanks for spotting it!

Sebastian-Wlo commented:
> Not much, couple lines. I'm fairly sure there's way to make that validation _automatic_ via model, but adding _manual_ check in the route function wouldn't be troublesome.

Would it require changing the content type back to FormData, or would JSON still work?
I'll add something like 'oldPassword' field for now, renaming them to match the updated model takes no time anyway

Sebastian-Wlo commented:
I'm not quite sure why, but I'm getting the 403 error (invalid password) when the 'old_password' is actually correct, or even if I'm just trying to change only the username.

gikf commented:
> I'm not quite sure why, but I'm getting the 403 error (invalid password) when the 'old_password' is actually correct, or even if I'm just trying to change only the username.
> 
> Edit: the latter is caused by the `old_password` being included - adding it to the request conditionally, or making name change and password change using separate forms would work. But I still can't figure out why passing in the correct password doesn't work.

Most likely I messed something up 😅 , I'll investigate.

Sebastian-Wlo commented:
> 
> Most likely I messed something up 😅 , I'll investigate.

I'm not so sure about that, I tend to find new and unexpected ways of getting wrong results:) But I would really appreciate it if you could take a look - I didn't see anything wrong with how #213 raises the error

gikf commented:
Password was being verified when `old_password` and `new_password` were non-`None`, but that can include when they were not touched in the form, and being the empty string.

Sebastian-Wlo commented:
> Password was being verified when `old_password` and `new_password` were non-`None`, but that can include when they were not touched in the form (being the empty string).

That fixed it, thanks!
I've changed the way the form asks for the old password and added a confirmation text (both are implemented in a really awkward way, but they kinda do the job) - not sure if the former is still needed now the Api ignores empty strings in password inputs, but it kinda makes it clear the old password is only needed for actually updating the passwords and not for the name change.

gikf commented:
> The header name is updated with the data that Api returns, and successful change to either username or password does return the whole user info - so it does "update", sort of... except, the name disappears if Api returns an error (since there's no `name` in the received data).

Not that header 😛! The one from the `Header` component, which takes user name from user state.

gikf commented on frontend/src/pages/MeEditPage.jsx:
```diff
@@ -1,10 +1,265 @@
+                  id='old-password'
+                  {...register('old_password', {
+                    minLength: 8,
+                    disabled: getValues('new_password') === '',
```
It's a bit weird, because sometimes the field gets enabled correctly, but usually not. It might be better to _watch_ the value changes to keep it more consistent.
https://react-hook-form.com/docs/useform/watch

Sebastian-Wlo replied:
Sorry! I must have forgotten to push the changes or something.
I removed that part line not long after your fix for the non-empty password fields. It was a temporary fix for changing the username when the Api was still expecting the `old_password` value for every submission.

Sebastian-Wlo commented:
> Not that header 😛! The one from the `Header` component, which takes user name from user state.

Right, __header__, not __heading__ (...I think I need a nap:D) - I'm currently trying to add that, still can't figure out what I should use to set the data so the name gets re-rendered when the form is submitted
